cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/ScheduleSpecificationCodec.java

```diff
@@ -32,22 +35,94 @@
  *
  */
 public class ScheduleSpecificationCodec extends AbstractSpecificationCodec<ScheduleSpecification> {
+
+  /**
+   * Schedule Type.
+   */
+  private enum ScheduleType {
+    /**
+     * Represents a {@link Schedule} object which, prior to 2.8.0, defines a time-based schedule.
+     * This one is never used for persistence
+     */
+    ORIGINAL_TIME,
+
+    /**
+     * Represents {@link TimeSchedule} objects.
+     */
+    TIME,
+
+    /**
+     * Represents {@link StreamSizeSchedule} objects.
+     */
+    STREAM;
+
+    private static ScheduleType fromSchedule(Schedule schedule) {
+      if (schedule instanceof StreamSizeSchedule) {
+        return STREAM;
+      } else if (schedule instanceof TimeSchedule) {
+        return TIME;
+      } else {
+        return ORIGINAL_TIME;
+      }
+    }
+  }
+
+  @SuppressWarnings("deprecation")
   @Override
   public JsonElement serialize(ScheduleSpecification src, Type typeOfSrc, JsonSerializationContext context) {
     JsonObject jsonObj = new JsonObject();
 
-    jsonObj.add("schedule", context.serialize(src.getSchedule(), Schedule.class));
+    ScheduleType scheduleType = ScheduleType.fromSchedule(src.getSchedule());
+    if (scheduleType.equals(ScheduleType.ORIGINAL_TIME)) {
```

Comment:
Ok that was to satisfy my test ``testBackwardCompatibility``. But the route that the test takes is not possible anymore, so I'll change it.